### ISSUE TITLE
Is it possible to add the neural fingerprint implementation?

### ISSUE THREAD

**Issue by simonfqy** [deepchem/deepchem#1160]
I was reading the `GraphConvTensorGraph` and related `classes`, and found that it is not from the paper named "Convolutional Networks on Graphs for Learning Molecular Fingerprints" as I originally expected. I found the design was proposed in "Low Data Drug Discovery with One-Shot Learning", but there were no comparisons as to whether this design is superior to the neural fingerprint. As it was not shown to have superior performance, I would suggest adding the original neural fingerprint model into our collection of models.

Is it possible for you to add this as a `featurizer` and `TensorGraph` model? I can also help you do it, but I cannot start until several weeks later, as my current project has a higher priority.
You can refer to [https://github.com/HIPS/neural-fingerprint](url) and [https://github.com/fllinares/neural_fingerprints_tf](url) for details.

**Comment by rbharath** [comment]
@simonfqy I believe that the fingerprints are very conceptually close. It's effectively the same implementation in practice with minor efficiency tweaks I think. Are there any particular differences that you think might make a difference here? If so, we'd welcome your help on duplicating.

**Comment by simonfqy** [comment]
@rbharath I didn't completely understand the implementation details. However, from what I understand, it seems that the neural fingerprint takes into account the higher-order neighbors of the atom under investigation within a graph distance specified by "radius", while `GraphConv` does not take into account of the higher-order neighbors when calculating the feature vector for each atom. This seems to be quite a difference to me.

**Comment by rbharath** [comment]
@lilleswing Would love your thoughts here. Does this match your recollection?

**Comment by lilleswing** [comment]
Our implementation and "neural fingerprints" are incredibly similar.  From their paper.

![selection_107](https://user-images.githubusercontent.com/1225605/37612438-a0e5d0ae-2b7b-11e8-9d92-b33024a39968.png)

The differences are...
line 9: We use a relu by default vs their "smooth" function.  I'm not sure what function they actually mean here
line 10: We do a batch-norm instead of a softmax.  Both are intended to get the atom feature vector to have certain ranges and distributions
line 11: we only do the sum of atom feature vectors to the molecule feature vector after all conv layers, not a running sum after each conv layer.

If we were to implement this model it would not be a "featurizer" it would be a Model.  @simonfqy the "higher-order neighbors" by specifying radius is the same as specifying number of layers into GraphConvTensorgraph class.

**Comment by simonfqy** [comment]
@lilleswing Thank you very much for your reply. Will find the time to implement it. Do you mean that the length of the list `graph_conv_layers` in `GraphConvTensorGraph` corresponds to the `radius` in neural fingerprint?

**Comment by lilleswing** [comment]
@simonfqy exactly correct.

**Comment by simonfqy** [comment]
@rbharath @lilleswing Can you please add a `Contribution Welcome` label to this issue? So that we could discuss the progress of the new feature if there are people other than me thinking about doing it themselves. Thanks!

**Comment by rbharath** [comment]
@simonfqy Done! I've marked this intermediate level since it will take some understanding of how to implement the neural fingerprint model precisely.

**Comment by rbharath** [comment]
I'm going to close this old issue since I don't think there's much demand for an exact implementation of the Duvenaud fingerprints versus our mild DeepChem variant. Please feel free to post here or re-open if there is interest!
